explainer.md

```diff
@@ -0,0 +1,195 @@
+    amount of data to be compressed is relatively small, this is not too
+    important. Applications involving huge volumes of data will probably still
+    want to use custom compression algorithms.
+*   Because of the ubiquity of these formats, they are useful in creation of
+    many file types, such as zip, pdf, svgz, and png.
+*   The latest bleeding-edge compression formats are likely to be soon
+    supplanted by new bleeding-edge compression formats. However, it is hard to
+    remove something from the web platform once added. In the worst case,
+    browser vendors could be left maintaining compression libraries that are no
+    longer supported upstream. For this reason, it is best to ship only
+    tried-and-tested algorithms.
+
+
+## Use cases
+
+*   Compressing data for upload.
+*   Compression and decompression for
+    *   Native files
+    *   Network protocols
+    *   In-memory databases
+*   Lazy decompression for downloads.
+
+
+## Example code
+
+### Gzip-compress a stream
+
+```javascript
+const compressedReadableStream = inputReadableStream.pipeThrough(new CompressionStream('gzip'));
+```
+
+### Deflate-compress an ArrayBuffer to a Uint8Array
+
+```javascript
+async function compressArrayBuffer(in) {
+  const cs = new CompressionStream('deflate');
+  const writer = cs.writable.getWriter();
+  writer.write(in);
+  writer.close();
+  const out = [];
+  const reader = cs.readable.getReader();
+  let totalSize = 0;
+  while (true) {
+    const { value, done } = await reader.read();
+    if (done)
+      break;
+    out.push(value);
+    totalSize += value.byteLength;
+  }
+  const concatenated = new Uint8Array(totalSize);
+  let offset = 0;
+  for (const array of out) {
+    concatenated.set(array, offset);
+    offset += array.byteLength;
+  }
+  return concatenated;
+}
+```
+
+### Gzip-decompress a Blob to a Blob
+
+This treats the input as a gzip file regardless of the mime-type. The output
+Blob has an empty mime-type.
+
+```javascript
+async function DecompressBlob(blob) {
+  const ds = new DecompressionStream('gzip');
+  const decompressedStream = blob.stream().pipeThrough(ds);
+  return await new Response(decompressedStream).blob();
+}
+```
+
+
+## End-user benefits
+
+Using this API, web developers can compress data to be uploaded, saving
+users time and bandwidth.
+
+As an alternative to this API, it is possible for web developers to bundle
+an implementation of a compression algorithm with their app. However, that
+would have to be downloaded as part of the app, costing the user time and
+bandwidth.
+
+
+## Considered alternatives
+
+*   Why not simply wrap the zlib API?
+
+    Not all platforms use zlib. Moreover, it is not a web-like API and
+    it’s hard to use. Implementing CompressionStream for zlib helps us
+    use it more easily.
+
+*   Why not support synchronous compression?
+
+    We want to be able to offload the work to another thread, which
```

Comment:
"We want" doesn't sound great.... why not phrase this is locking the main thread for potentially a significant amount of time, leading to jank and a bad user experience for users (similar to sync XHR).